File: src/mtc_pipeline/src/stages/pick_place_stages.cpp
#include "mtc_pipeline/pick_place_stages.hpp"
#include "mtc_pipeline/gripper_utils.hpp"

#include <moveit/task_constructor/stages/move_to.h>
#include <moveit_msgs/msg/constraints.hpp>
#include <moveit_msgs/msg/joint_constraint.hpp>

// Note: Helper functions moved to eliminate duplication:
//   - Gripper utilities -> gripper_utils.hpp
//   - Wrist constraint -> BaseStages::create_wrist3_level_constraint()

PickPlaceStages::PickPlaceStages(const rclcpp::Node::SharedPtr& node)
    : BaseStages(node) {}

std::unique_ptr<mtc::Stage> PickPlaceStages::make_move_to_named_stage(
    const std::string& label,
    const std::string& pose_key,
    const nlohmann::json& poses,
    const mtc::solvers::PlannerInterfacePtr& planner)
{
    const auto& joint_pose = poses.at(pose_key);
    if (!joint_pose.is_array() || joint_pose.size() != 6) {
        RCLCPP_ERROR(node()->get_logger(), "'%s' must be array of 6 joint angles", pose_key.c_str());
        return nullptr;
    }

    auto stage = std::make_unique<mtc::stages::MoveTo>(label, planner);
    stage->properties().configureInitFrom(mtc::Stage::PARENT, {"group", "ik_frame"});
    stage->setGroup(default_arm_group_name());
    stage->setGoal(joints_from_degrees(joint_pose.get<std::vector<double>>()));
    return stage;
}

std::unique_ptr<mtc::Stage> PickPlaceStages::make_gripper_stage(
    const std::string& label,
    const mtc::solvers::PlannerInterfacePtr& planner,
    bool open,
    const std::string& gripper_type)
{
    auto stage = std::make_unique<mtc::stages::MoveTo>(label, planner);
    stage->setGroup(mtc_pipeline::gripper_utils::get_group_name(gripper_type));
    stage->setGoal(mtc_pipeline::gripper_utils::get_state_name(gripper_type, open));
    return stage;
}

bool PickPlaceStages::run(const mtc_pipeline::action::PickPlaceAction::Goal& goal)
{
    nlohmann::json poses;
    try {
        poses = nlohmann::json::parse(goal.poses_json);
    } catch (const nlohmann::json::exception& e) {
        RCLCPP_ERROR(node()->get_logger(), "Failed to parse poses_json: %s", e.what());
        return false;
    }

    auto task = create_task_template("Pick and Place");
    auto pipeline = make_pipeline_planner();
    auto gripper_planner = make_joint_interpolation_planner();

    // Helper to add constrained move stage
    auto add_constrained_move = [&](const std::string& label, const std::string& pose_key) {
        auto stage = make_move_to_named_stage(label, pose_key, poses, pipeline);
        if (!stage) return false;
        if (auto* move = dynamic_cast<mtc::stages::MoveTo*>(stage.get())) {
            move->setPathConstraints(create_wrist3_level_constraint());
        }
        task.add(std::move(stage));
        return true;
    };

    // Pick sequence
    task.add(make_gripper_stage("open gripper", gripper_planner, true, goal.gripper));
    if (!add_constrained_move("pickup approach", goal.pick_approach)) return false;
    if (!add_constrained_move("pickup", goal.pick_target)) return false;
    task.add(make_gripper_stage("close gripper", gripper_planner, false, goal.gripper));
    if (!add_constrained_move("pickup retreat", goal.pick_approach)) return false;

    // Place sequence (no wrist constraint - more flexibility needed)
    auto add_move = [&](const std::string& label, const std::string& pose_key) {
        auto stage = make_move_to_named_stage(label, pose_key, poses, pipeline);
        if (!stage) return false;
        task.add(std::move(stage));
        return true;
    };

    if (!add_move("place approach", goal.place_approach)) return false;
    if (!add_move("place", goal.place_target)) return false;
    task.add(make_gripper_stage("release", gripper_planner, true, goal.gripper));
    if (!add_move("place retreat", goal.place_approach)) return false;

    // Return home
    auto home = std::make_unique<mtc::stages::MoveTo>("return home", pipeline);
    home->properties().configureInitFrom(mtc::Stage::PARENT, {"group", "ik_frame"});
    home->setGroup(default_arm_group_name());
    home->setGoal("moveit_home");
    task.add(std::move(home));

    return load_plan_execute(task);
}

File: src/mtc_py/mtc_py_lib/stages/pick_place_stages.py
"""PickPlace stages - Python equivalent of pick_place_stages.hpp/cpp.

Pick and place sequence: approach -> grip -> retreat -> approach -> release -> retreat.
Uses constrained motion during pick to maintain tool orientation.
"""

from typing import Any, Dict, Optional
from moveit.task_constructor import core, stages
from mtc_py_lib.stages.base_stages import BaseStages
from mtc_py_lib.utils.gripper_utils import get_group_name, get_state_name


class PickPlaceStages(BaseStages):
    """Handles pick and place sequences with gripper operations."""

    def run(self, goal) -> bool:
        """Execute PickPlace action.

        Sequence:
        1. Open gripper
        2. Move to pick approach (constrained)
        3. Move to pick target (constrained)
        4. Close gripper
        5. Move to pick approach (retreat, constrained)
        6. Move to place approach
        7. Move to place target
        8. Open gripper (release)
        9. Move to place approach (retreat)
        10. Return home

        Args:
            goal: PickPlaceAction.Goal with fields:
                - gripper: "hande" or "epick"
                - pick_approach: Approach pose name
                - pick_target: Grasp pose name
                - place_approach: Place approach pose name
                - place_target: Place target pose name
                - poses_json: JSON string with pose definitions

        Returns:
            True if successful, False otherwise
        """
        # Parse poses (required for pick/place operations)
        poses = self.parse_poses(goal.poses_json, required=True)
        if poses is None:
            return False

        task = self.create_task_template("Pick and Place")
        pipeline = self.make_pipeline_planner()
        gripper_planner = self.make_joint_interpolation_planner()

        # === PICK SEQUENCE ===
        # 1. Open gripper
        open_stage = self._make_gripper_stage(
            "open gripper", gripper_planner, True, goal.gripper
        )
        if open_stage:
            task.add(open_stage)

        # 2. Move to pick approach (constrained)
        stage = self._make_constrained_move_stage(
            "pickup approach", goal.pick_approach, poses, pipeline
        )
        if not stage:
            return False
        task.add(stage)

        # 3. Move to pick target (constrained)
        stage = self._make_constrained_move_stage(
            "pickup", goal.pick_target, poses, pipeline
        )
        if not stage:
            return False
        task.add(stage)

        # 4. Close gripper
        close_stage = self._make_gripper_stage(
            "close gripper", gripper_planner, False, goal.gripper
        )
        if close_stage:
            task.add(close_stage)

        # 5. Retreat from pick (constrained)
        stage = self._make_constrained_move_stage(
            "pickup retreat", goal.pick_approach, poses, pipeline
        )
        if not stage:
            return False
        task.add(stage)

        # === PLACE SEQUENCE (no wrist constraint - more flexibility needed) ===
        # 6. Move to place approach
        stage = self._make_move_to_named_stage(
            "place approach", goal.place_approach, poses, pipeline
        )
        if not stage:
            return False
        task.add(stage)

        # 7. Move to place target
        stage = self._make_move_to_named_stage(
            "place", goal.place_target, poses, pipeline
        )
        if not stage:
            return False
        task.add(stage)

        # 8. Release (open gripper)
        release_stage = self._make_gripper_stage(
            "release", gripper_planner, True, goal.gripper
        )
        if release_stage:
            task.add(release_stage)

        # 9. Retreat from place
        stage = self._make_move_to_named_stage(
            "place retreat", goal.place_approach, poses, pipeline
        )
        if not stage:
            return False
        task.add(stage)

        # 10. Return home
        home = stages.MoveTo("return home", pipeline)
        home.group = self.arm_group
        self._set_ik_frame(home)
        home.setGoal("moveit_home")
        task.add(home)

        return self.load_plan_execute(task)

    def _make_move_to_named_stage(
        self,
        label: str,
        pose_key: str,
        poses: Dict[str, Any],
        planner
    ) -> Optional[stages.MoveTo]:
        """Create a MoveTo stage for a named joint pose.

        Args:
            label: Stage name
            pose_key: Key in poses dict
            poses: Dictionary of pose definitions
            planner: Planner to use

        Returns:
            Configured MoveTo stage, or None on error
        """
        joint_pose = self.get_joint_pose(poses, pose_key)
        if joint_pose is None:
            return None

        stage = stages.MoveTo(label, planner)
        stage.group = self.arm_group
        self._set_ik_frame(stage)
        stage.setGoal(self.joints_from_degrees(joint_pose))
        return stage

    def _make_constrained_move_stage(
        self,
        label: str,
        pose_key: str,
        poses: Dict[str, Any],
        planner
    ) -> Optional[stages.MoveTo]:
        """Create a MoveTo stage with wrist3 constraint.

        Args:
            label: Stage name
            pose_key: Key in poses dict
            poses: Dictionary of pose definitions
            planner: Planner to use

        Returns:
            Configured MoveTo stage with path constraints, or None on error
        """
        stage = self._make_move_to_named_stage(label, pose_key, poses, planner)
        if stage:
            stage.setPathConstraints(self.create_wrist3_level_constraint())
        return stage

    def _make_gripper_stage(
        self,
        label: str,
        planner,
        open_gripper: bool,
        gripper_type: str
    ) -> Optional[stages.MoveTo]:
        """Create a gripper open/close stage.

        Args:
            label: Stage name
            planner: Planner to use
            open_gripper: True to open, False to close
            gripper_type: "hande" or "epick"

        Returns:
            Configured MoveTo stage for gripper, or None if no gripper
        """
        gripper_group = get_group_name(gripper_type)
        if not gripper_group:
            self.logger.info(f"No gripper for type '{gripper_type}' - skipping")
            return None

        stage = stages.MoveTo(label, planner)
        stage.group = gripper_group
        stage.setGoal(get_state_name(gripper_type, open_gripper))
        return stage
